### PR TITLE
Rec: shared and sharded negcache

### PR DESCRIPTION
**Pull request by omoerbeek**
### Short description

While the main record cache is shared between threads since some time, the negcache was not. This makes is so.

While there, simplify the `lruReplacingInsert()` result to just signal if an entry was inserted or replaced (this makes counting more easy) and generalize the invalidating of the per-shard cache used in the record cache so the code can be more easily be reused in the negcache.

This code keeps track of mutex acquired/contended counts. Don't know if this is going to stay, it is not exported in any way atm.

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [X] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/pdns_recursor.cc [hunk 1]
```diff
@@ -131,6 +131,7 @@ static thread_local uint64_t t_frameStreamServersGeneration;
 
 thread_local std::unique_ptr<MT_t> MT; // the big MTasker
 std::unique_ptr<MemRecursorCache> s_RC;
+std::unique_ptr<NegCache> s_negcache;
```
Comment (rgacogne) [inline pdns/pdns_recursor.cc:134]: The `s_` prefix is usually for variables static to a file, or static to a class, so perhaps use `g_` instead?

File: pdns/recursordist/negcache.cc [hunk 2]
```diff
@@ -70,24 +101,27 @@ bool NegCache::getRootNXTrust(const DNSName& qname, const struct timeval& now, N
  */
 bool NegCache::get(const DNSName& qname, const QType& qtype, const struct timeval& now, NegCacheEntry& ne, bool typeMustMatch)
 {
-  const auto& idx = d_negcache.get<2>();
+  auto& map = getMap(qname);
+  const lock l(map);
+
+  const auto& idx = map.d_map.get<2>();
```
Comment (rgacogne) [inline pdns/recursordist/negcache.cc:107]:
```suggestion
  const auto& idx = map.d_map.get<NegCacheEntry>();
```

File: pdns/recursordist/negcache.cc [hunk 3]
```diff
@@ -70,24 +101,27 @@ bool NegCache::getRootNXTrust(const DNSName& qname, const struct timeval& now, N
...
   auto range = idx.equal_range(qname);
   auto ni = range.first;
 
   while (ni != range.second) {
     // We have an entry
     if ((!typeMustMatch && ni->d_qtype.getCode() == 0) || ni->d_qtype == qtype) {
       // We match the QType or the whole name is denied
-      auto firstIndexIterator = d_negcache.project<0>(ni);
+      auto firstIndexIterator = map.d_map.project<0>(ni);
```
Comment (rgacogne) [inline pdns/recursordist/negcache.cc:115]:
```suggestion
      auto firstIndexIterator = map.d_map.project<CompositeKey>(ni);
```

**Review by rgacogne — APPROVED** [review]

File: pdns/pdns_recursor.cc (hunk 1 above)
Comment (omoerbeek) [reply pdns/pdns_recursor.cc:134]: Right. I opt for` g_NegCache`  and might rename `s_RC`  then into `g_RecCache` as well.

**Pull request merged**